src/main/java/eu/dissco/core/digitalspecimenprocessor/service/ProcessingService.java

```diff
@@ -1,632 +1,417 @@
-              tuple.digitalSpecimenEvent().digitalMediaEvents(),
-              tuple.digitalMediaProcessResult());
-        }
-    ).collect(Collectors.toSet());
+    if (!mediaProcessResult.changedDigitalMedia().isEmpty()) {
+      results.addAll(
+          digitalMediaService.updateExistingDigitalMedia(mediaProcessResult.changedDigitalMedia(),
+              pidProcessResults));
+    }
+    return results;
   }
 
-  private boolean updateHandles(List<UpdatedDigitalSpecimenTuple> updatedDigitalSpecimenTuples) {
-    var digitalSpecimensToUpdate = updatedDigitalSpecimenTuples.stream()
-        .filter(tuple -> fdoRecordService.handleNeedsUpdate(
-            tuple.currentSpecimen().digitalSpecimenWrapper(),
-            tuple.digitalSpecimenEvent().digitalSpecimenWrapper()))
-        .toList();
 
-    if (!digitalSpecimensToUpdate.isEmpty()) {
-      try {
-        log.info("Updating {} doi records", digitalSpecimensToUpdate.size());
-        var requests = fdoRecordService.buildUpdateHandleRequest(digitalSpecimensToUpdate);
-        handleComponent.updateHandle(requests);
-      } catch (PidException e) {
-        log.error("Unable to update DOI record. Not proceeding with update. ", e);
-        try {
-          for (var tuple : updatedDigitalSpecimenTuples) {
-            publisherService.deadLetterEvent(tuple.digitalSpecimenEvent());
+  private Set<DigitalSpecimenEvent> removeDuplicateSpecimensInBatch(
+      List<DigitalSpecimenEvent> events) {
+    var uniqueSet = new LinkedHashSet<DigitalSpecimenEvent>();
+    var map = events.stream()
+        .collect(
+            Collectors.groupingBy(event -> event.digitalSpecimenWrapper().physicalSpecimenID()));
+    for (Entry<String, List<DigitalSpecimenEvent>> entry : map.entrySet()) {
+      if (entry.getValue().size() > 1) {
+        log.warn("Found {} duplicate specimen in batch for id {}", entry.getValue().size(),
+            entry.getKey());
+        for (int i = 0; i < entry.getValue().size(); i++) {
+          if (i == 0) {
+            uniqueSet.add(entry.getValue().get(i));
+          } else {
+            republishSpecimenEvent(entry.getValue().get(i));
           }
-        } catch (JsonProcessingException jsonEx) {
-          log.error(DLQ_FAILED, updatedDigitalSpecimenTuples, jsonEx);
         }
-        return false;
+      } else {
+        uniqueSet.add(entry.getValue().get(0));
       }
     }
-    return true;
+    return uniqueSet;
   }
 
-  private boolean publishUpdateEvent(UpdatedDigitalSpecimenRecord updatedDigitalSpecimenRecord) {
-    try {
-      publisherService.publishUpdateEvent(updatedDigitalSpecimenRecord.digitalSpecimenRecord(),
-          updatedDigitalSpecimenRecord.jsonPatch());
-      return true;
-    } catch (JsonProcessingException e) {
-      log.error("Failed to publish update event", e);
-      return false;
-    }
+  private Set<DigitalMediaEvent> removeDuplicateMediaInBatch(
```

Comment:
Just to check, we can deduplicate here without republishing because we removed the coupling.

Reply:
Exactly. If 3 specimens point to the same media, it would be pretty redundant to republish them. It's better to publish that media object at once, associated with all 3 specimens at the onset